db/view/view.cc

```diff
@@ -2645,6 +2646,30 @@ update_backlog node_update_backlog::fetch() {
     return std::max(fetch_shard(this_shard_id()), _max.load(std::memory_order_relaxed));
 }
 
+future<std::optional<update_backlog>> node_update_backlog::fetch_if_changed(std::optional<db::view::update_backlog> last_backlog) {
+    auto np = co_await map_reduce(boost::irange(0u, smp::count),
+            [this] (shard_id shard) {
+                return smp::submit_to(shard, [this, shard] {
+                    if (_backlogs[shard].need_publishing) {
+                        _backlogs[shard].need_publishing = need_publishing::no;
+                        return need_publishing::yes;
+                    }
+                    return need_publishing::no;
+                });
+            },
+            need_publishing::no,
+            [] (need_publishing a, need_publishing b) {
+                return a || b;
+            });
+
+    auto max = fetch();
```

Comment:
This will be prone to races. Consider the following scenario:

- `_backlog` is updated on shard 1, the `need_publishing` flag is set,
- The `_last_update` is recalculated, next one will happen 10ms from now or later,
- 5ms pass,
- `fetch_if_changed` and clears the `need_publishing` flag on shard 1, but uses `fetch` to read the old value,
- There are no MV writes anymore and we didn't publish the updated backlog on shard 1.

You should collect the backlogs in the `submit_to` lambda and calculate the maximum in the reducer lambda (you should still collect the np flag in addition to this).

Reply:
Added in the rebase. This method now updates the max backlog unconditionally (disregarding the `10ms` interval - it's only run once every second anyway), and returns the new max only if the `need_publishing` flag was set somewhere